### PR TITLE
Hooks for VM migration plans

### PR DESCRIPTION
**Pull request by RichardHoch**
MTV 2.2+ (target: 2.3.0)

Resolves: https://issues.redhat.com/browse/MTV-246

Preview: https://file.emea.redhat.com/rhoch/hook/html-single/#adding_hooks_to_a_migration_plan

### CODE REVIEW

**Comment by fbladilo** [comment]
@jmontleon As far as I can tell we don't have a specific MTV hook-runner image for downstream and all of ALM examples point to konveyor hook runner image but I will let @jortel have final say on this one. Potentially we could re-use the MTC image and update our examples downstream, assuming MTC's downstream image is identical to konveyor's one.

**Review by jortel — APPROVED** [review]
LGTM

File: documentation/modules/about-hooks-for-migration-plans-api.adoc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+:_content-type: CONCEPT
+[id="about-hooks-for-migration-plans-api_{context}"]
+= API-based hooks for {project-short} migration plans
+
+You can add hooks to a migration plan from the command line by using the {project-full} API.
+
+[discrete]
+== Default hook image
+The default hook image for an {project-short} hook is `quay.io/konveyor/hook-runner:latest`. The image is based on the Ansible Runner image with the addition of `python-openshift` to provide Ansible Kubernetes resources and a recent `oc` binary.
+// The most recent?
```
Comment (ahadas) [inline documentation/modules/about-hooks-for-migration-plans-api.adoc:14]: I see that quay.io/konveyor/hook-runner keeps being updated but it's not that great that we refer to an image on quay.io (we can but we also need to explain how to change this image, for disconnected environments) and specifically to an image in `konveyor`

Comment (ahadas) [reply documentation/modules/about-hooks-for-migration-plans-api.adoc:14]: @liranr23 can you please take a look at productizing the hook-runner image?

File: documentation/modules/adding-hooks-to-migration-plans-api.adoc [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      state: latest
+
+  - name: Create Test File
+    copy:
+      dest: /test.txt
+      content: "Hello World"
+      mode: 0644
+----
+
+. Create the plan using the hook
```
Comment (ahadas) [inline documentation/modules/adding-hooks-to-migration-plans-api.adoc:146]: Create a plan that uses the hook?

File: documentation/modules/about-hooks-for-migration-plans-api.adoc (hunk 1 above)
Comment (liranr23) [reply documentation/modules/about-hooks-for-migration-plans-api.adoc:14]: let's use `registry.redhat.io/rhmtc/openshift-migration-hook-runner-rhel8:v1.8.2-2`

**Comment by RichardHoch** [comment]
@liranr23 @ahadas I made the changes you suggested. Please review again.

File: documentation/modules/adding-hooks-to-migration-plans-api.adoc [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
++
+[source,yaml]
+----
+apiVersion: forklift.konveyor.io/v1beta1
+kind: Hook
+metadata:
+  name: playbook
+  namespace: konveyor-forklift
+spec:
+  image: quay.io/konveyor/hook-runner
```
Comment (liranr23) [inline documentation/modules/adding-hooks-to-migration-plans-api.adoc:65]: might need to change this image as well

Comment (RichardHoch) [reply documentation/modules/adding-hooks-to-migration-plans-api.adoc]: Done.

**Comment by RichardHoch** [comment]
@liranr23 @ahadas I made the changes you suggested. Please review again.

**Review by liranr23 — APPROVED** [review]

**Comment by RichardHoch** [comment]
@qiyuann @ahadas Please review this PR.

File: documentation/modules/about-hooks-for-migration-plans-api.adoc [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+[discrete]
+== Hook execution
+An Ansible playbook that is provided as part of a migration hook is mounted into the hook container as a `ConfigMap`. The hook container is run as a job on the desired cluster, using the default `ServiceAccount` in the `konveyor-forklift` namespace.
+
+[discrete]
+== PreHooks and PostHooks
+You specify hooks per VM and you can run each as a _PreHook_ or a _PostHook_. In this context, a PreHook is a hook that is run before a migration and a PostHook is a hook that is run after a migration.
+
+When you adding a hook, you must specify the namespace where the hook CR is located, the name of the hook, and specify whether the hook is a PreHook or PostHook.
```
Comment (ahadas) [inline documentation/modules/about-hooks-for-migration-plans-api.adoc:23]:
```suggestion
When you add a hook, you must specify the namespace where the hook CR is located, the name of the hook, and specify whether the hook is a PreHook or PostHook.
```

File: documentation/modules/adding-hooks-to-migration-plans-api.adoc [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+:_content-type: PROCEDURE
+[id="migration-migrating-applications-api_{context}"]
+= Adding Hook CRs to a VM migration by using the {project-short} API
+
+You can add a `PreHook` or a `PostHook` Hook CR when you migrate an application from the command line by using the {project-full} API. A `PreHook` runs before a migration, a `PostHook`, after.
+
+[NOTE]
+====
+You can retrieve additional information stored in secrets or a `configMap`s by using `k8s` modules.
```
Comment (ahadas) [inline documentation/modules/adding-hooks-to-migration-plans-api.adoc:13]: "or a `configMap`s"? did you mean "or qconfigMaps`?

File: documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc [hunk 6]
```diff
@@ -220,6 +220,12 @@ include::modules/canceling-migration-cli.adoc[leveloffset=+2]
 
 include::modules/changing-precopy-intervals.adoc[leveloffset=+2]
 
+=== Adding hooks to a migration plan
+
+include::modules/about-hooks-for-migration-plans-api.adoc[leveloffset=+3]
+
+include::modules/adding-hooks-to-migration-plans-api.adoc[leveloffset=+3]
```
Comment (ahadas) [inline documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc:227]: the name of the files are not that good..

**Comment by RichardHoch** [comment]
@ahadas I made the changes you suggested. Please review this PR again.

File: documentation/modules/adding-hook-crs-to-migration-plans-via-api.adoc [hunk 7]
```diff
@@ -0,0 +1,179 @@
+// Module included in the following assemblies:
+//
+// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc
+
+:_content-type: PROCEDURE
+[id="adding-hook-crs-migration-plans-api_{context}"]
+= Adding Hook CRs to a VM migration by using the {project-short} API
+
+You can add a `PreHook` or a `PostHook` Hook CR when you migrate an application from the command line by using the {project-full} API. A `PreHook` runs before a migration, a `PostHook`, after.
```
Comment (ahadas) [inline documentation/modules/adding-hook-crs-to-migration-plans-via-api.adoc:9]:
```suggestion
You can add a `PreHook` or a `PostHook` Hook CR when you migrate a virtual machine from the command line by using the {project-full} API. A `PreHook` runs before a migration, a `PostHook`, after.
```

File: documentation/modules/about-api-hooks-in-migration-plansi.adoc
Comment (ahadas) [inline documentation/modules/about-api-hooks-in-migration-plansi.adoc:1]: the name of the file should rather be documentation/modules/about-api-hooks-in-migration-plans.adoc

Comment (RichardHoch) [reply documentation/modules/about-api-hooks-in-migration-plansi.adoc:1]: Actually, that file was a unneeded copy of documentation/modules/about-api-hooks-in-migration-plans.adoc, so i deleted it.

**Comment by RichardHoch** [comment]
@ahadas I made the changes you suggested. Please review again.

**Comment by ahadas** [comment]
> @ahadas I made the changes you suggested. Please review again.

@RichardHoch it seems like you didn't upload the changes

**Comment by RichardHoch** [comment]
@ahadas Pushed the commit; please review again.

**Comment by RichardHoch** [comment]
@ahadas That was a "dummy" file I used while fixing the merge conflict, so I have deleted it. The file was not called by master.adoc, but thanks for catching that. No need for clutter.

**Pull request merged**